scripts/generate-CHANGES.py

```diff
@@ -0,0 +1,324 @@
+	if p.returncode != 0:
+		raise Exception('"git log" failed: %s', stderr)
+	return stdout
+
+def main():
+	log = gitLogOutput()
+
+	# The authors dict maps between email addresses
+	# and author names.
+	authors = {}
+
+	# The dates dictionary maps from
+	# dates, to another dictionary
+	# that maps from email addresses
+	# to a sorted list of commit subject
+	# lines.
+	#
+	# In pseudo-code, the type of 'dates' is:
+	# 
+	#     dict<string(YYMMDD), dict<string(email), list(subject)>>
+	#
+	dates = {}
+
+	# lastSubject contains the last subject line
+	# processed. It is used to detect duplicate
+	# subject lines and discard them.
+	lastSubject = ""
+
+	for line in log.split('\n'):
+		commit, author, email, date, subject = line.split('\0')
+
+		if author == "Thorvald Natvig":
+			# Because someone forgets to set their git username on every platform...
+			email = "slicer@users.sourceforge.net"
+
+		# Back in the old days, we appended [$SourceForgeUserName]
+		# in our commit messages when we applied patches by contributors
+		# on SourceForge.
+		#
+		# For example:
+		#
+		#     Half-Life 2 Deathmatch Positional Audio [Snares]
+		#
+		# The code below extracts author names from the subject line
+		# of the commit message, and updates the 'author' and
+		# 'email' variables to point to the SourceForge user
+		# specified in the square brackets. The email address
+		# is simply the lowercase username @users.sourceforge.net.
+		# 
+		patchAuthorRe = "\s*\[([\w-]+)\]\s*$"
+		match = re.search(patchAuthorRe, subject, flags=re.UNICODE)
+		if match is not None:
+			author = match.groups()[0]
+			email = "{0}@users.sourceforge.net".format(author.lower())
+
+			# Remove the [$SourceForgeUserName] from the subject line.
+			subject = re.sub(patchAuthorRe, "", subject)
+
+		# Record the first occurrence of an author's
+		# email address and name in the author <-> email map.
+		#
+		# This mapping is used 
+		if authors.get(email, None) is None:
+			authors[email] = author
+
+		if matchRegexp("^Merge branch 'master", subject):
+			continue
+		if matchRegexp("^Indent and submodule update", subject):
+			continue
+		if matchRegexp("^Indent, changelog and submodule update", subject):
+			continue
+		if matchRegexp("^Indent, changelog, submodule and language update", subject):
+			continue
+		if matchRegexp("^Indent run", subject):
+			continue
+		if matchRegexp("^\*\*\* empty log message", subject):
+			continue
+		if matchRegexp("^git-svn-id: http", subject):
+			continue
+		if matchRegexp("^TEST", subject):
+			continue
+		if matchRegexp("^Merge pull request", subject):
+			continue
+
+		# Skip duplicate subject lines.
+		if subject == lastSubject:
+			continue
+		lastSubject = subject
+
+		# Format the commit hash and subject line.
+		entry = perlCompatibleTextWrap(subject, 76, initial_indent="    {0}  ".format(commit), subsequent_indent="\t     ")
+
+		# Add the entry to the dictionary structure.
+		dateDict = dates.get(date, {})
```

Comment:
Would it get much cleaner?
We'll still have nested dictionaries that need to be set up, no?
Or can I nest defaultdicts?